### PR TITLE
Grouped pad/shift/move operations

### PR DESCRIPTION
**Pull request by madamczyk-intel**
Two main changes:
* reworks main components of concat so that they're run in bulk to minimize host overhead.
* left shifting is done with a single biggest chunk size

### CODE REVIEW

**Comment by madamczyk-intel** [comment]
@kdamaszk please review

File: server/text_generation_server/models/causal_lm.py [hunk 1]
```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
...
-        tensor = torch.roll(tensor, c, dim)
-        htorch.core.mark_step()
-    return tensor
+def biggest_single_chunk(offset):
+    if offset < 0:
+        idx = bisect.bisect(CHUNK_SIZES, -offset)
+        return -CHUNK_SIZES[idx-1]
+    elif offset > 0:
+        idx = bisect.bisect(CHUNK_SIZES, offset)
+        candidate = CHUNK_SIZES[idx]
```
Comment (madamczyk-intel) [inline server/text_generation_server/models/causal_lm.py]: Dead code

File: server/text_generation_server/models/causal_lm.py [hunk 2]
```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
...
+        return 0
 
 
-def shift(tensor, dim, offset):
-    assert dim < 0, 'Only negative dims are supported'
-    if offset == 0:
+def grouped_pad(tensor_groups, dims, values):
+    paddings = [MAX_TOTAL_TOKENS - tensors[0].size(dim) if dim is not None else 0 for tensors, dim in zip(tensor_groups, dims)]
+    grouped_result = []
+    for tensors, dim, padding, value in zip(tensor_groups, dims, paddings, values):
```
Comment (madamczyk-intel) [inline server/text_generation_server/models/causal_lm.py]: add assert for dim=-1 or dim=-2

File: server/text_generation_server/models/causal_lm.py [hunk 3]
```diff
@@ -218,6 +277,99 @@ def to_pb(self) -> generate_pb2.CachedBatch:
...
+
+    def free_indices_generator(self):
+        used = set(req.idx for req in self.requests)
+        return (i for i in range(self.batch_size) if i not in used)
+
+    def move_data(self, src_batches):
+        dst_tensors, _, dst_dims = self.get_tensor_groups()
+        free_indices_gen = self.free_indices_generator()
+        for src_b in src_batches:
+            #src_indices = src_b.used_indices()
```
Comment (madamczyk-intel) [inline server/text_generation_server/models/causal_lm.py]: dead code

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py]: `candidate` is not used anywhere

File: server/text_generation_server/models/causal_lm.py [hunk 4]
```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
...
+def biggest_single_chunk(offset):
+    if offset < 0:
+        idx = bisect.bisect(CHUNK_SIZES, -offset)
+        return -CHUNK_SIZES[idx-1]
+    elif offset > 0:
+        idx = bisect.bisect(CHUNK_SIZES, offset)
+        candidate = CHUNK_SIZES[idx]
+        return CHUNK_SIZES[idx-1]
+    else:
+        return 0
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py:99]: since `candidate is not needed, wouldn't be better to use abs and copysign?
e.g.:
```python
def biggest_single_chunk(offset):
    if offset != 0:
        idx = bisect.bisect(CHUNK_SIZES, abs(offset))
        return -math.copysign(CHUNK_SIZES[idx-1], offset)
    else:
        return 0
`

File: server/text_generation_server/models/causal_lm.py [hunk 5]
```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
...
 
 
-def shift(tensor, dim, offset):
-    assert dim < 0, 'Only negative dims are supported'
-    if offset == 0:
+def grouped_pad(tensor_groups, dims, values):
+    paddings = [MAX_TOTAL_TOKENS - tensors[0].size(dim) if dim is not None else 0 for tensors, dim in zip(tensor_groups, dims)]
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py]: you iterate over it twice, here and in line 108. I guess you could just simply calculate padding in line 109 as:
`padding = MAX_TOTAL_TOKENS - tensors[0].size(dim) if dim is not None else 0`
both `tensors` and `dim` is available on the loop at line 108.
It will save some host time to trace it

File: server/text_generation_server/models/causal_lm.py [hunk 6]
```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
...
+
+
+def roll(tensor, chunk, dim, merge_graphs):
+    if dim is None:
         return tensor
-    chunks = generate_shift_chunks(offset)
-    tensor = roll(tensor, dim, chunks)
+    tensor = torch.roll(tensor, chunk, dim)
+    if not merge_graphs:
+        htorch.core.mark_step()
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py:122]: can you explain why mark step only when merge graphs?

File: server/text_generation_server/models/causal_lm.py [hunk 7]
```diff
@@ -218,6 +277,99 @@ def to_pb(self) -> generate_pb2.CachedBatch:
...
+            #src_indices = src_b.used_indices()
+            src_indices = to_tensor_indices(src_b.used_indices(), self.input_ids.device)
+            dst_indices = to_tensor_indices(src_b.update_indices(free_indices_gen), self.input_ids.device)
+            src_tensors, _, src_dims = src_b.get_tensor_groups()
+            # move past_key
+            dst_tensors[3:4] = grouped_move(dst_tensors[3:4], dst_dims[3:4], dst_indices, src_tensors[3:4], src_dims[3:4], src_indices)
+            # move past_value
+            dst_tensors[4:5] = grouped_move(dst_tensors[4:5], dst_dims[4:5], dst_indices, src_tensors[4:5], src_dims[4:5], src_indices)
+            # move rest
+            dst_tensors[:3] = grouped_move(dst_tensors[:3], dst_dims[:3], dst_indices, src_tensors[:3], src_dims[:3], src_indices)
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py:366]: what `3:4` `4:5` and `:3` indices mean?

**Review by mswiniarsk — CHANGES_REQUESTED** [review]

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py]: I've already removed it. Please recheck on latest commit.

File: server/text_generation_server/models/causal_lm.py (hunk 4 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:99]: Nice! I didn't know copysign! It returns a float so a cast would be needed but still it would be better.

File: server/text_generation_server/models/causal_lm.py (hunk 5 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py]: Good idea! Thanks!

File: server/text_generation_server/models/causal_lm.py (hunk 6 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:122]: So it comes down to this:
if merge_graphs -> we add a mark_step in grouped_roll
if not merge_graphs -> we add a mark_step in roll

This way we either have a single graph with N tensors, or we have N graphs with a single tensor roll. This is important for left_shifting as the tensors are big and doing it in bulk significantly increases memory usage.

File: server/text_generation_server/models/causal_lm.py (hunk 7 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:366]: Both dst_tensors and src_tensors are tensor groups returned by get_tensor_groups which returns:
`[[self.input_ids], [self.attention_mask], [self.position_ids], past_keys, past_values]`
`list[3:4]` is used here instead of `[list[3]]` because grouped_* operations assume the inputs are grouped. In the end we're first moving the past_keys group ([3:4]) then the past_value group ([4:5]) followed by small tensors like input_ids, attention_mask and position_ids

Comment (mswiniarsk) [reply server/text_generation_server/models/causal_lm.py:366]: can you please elaborate on comment about this? So it would be well understood by future developers

File: server/text_generation_server/models/causal_lm.py [hunk 8]
```diff
@@ -152,13 +205,17 @@ def forward(*args, **kwargs):
...
-            htorch.core.mark_step()
-    return tensors
+def hpu_graph_fn(fn):
+    class FnModule(torch.nn.Module):
+        def forward(self, *args, **kwargs):
+            return fn(*args, **kwargs)
+    return wrap_in_hpu_graph(FnModule(), disable_tensor_cache=False)
+
+
+def enable_hpu_graphs():
```
Comment (kdamaszk) [inline server/text_generation_server/models/causal_lm.py]: This function name is a little bit confusing -- probably 'enable_hpu_graphs_on_grouped_move' would be better. BTW can we remove this from that PR as it is not functional?

File: server/text_generation_server/models/causal_lm.py [hunk 9]
```diff
@@ -593,6 +670,9 @@ def __init__(
                 model = wrap_in_hpu_graph(model, disable_tensor_cache=True)
         model = self.setup_quantization(model)
 
+        if self.enable_hpu_graph:
+            enable_hpu_graphs()
```
Comment (kdamaszk) [inline server/text_generation_server/models/causal_lm.py]: This is also confusing.. We are enabling HPU graphs 3 lines above, so this function name is not accurate.

File: server/text_generation_server/models/causal_lm.py (hunk 8 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py]: Ok, let's remove it for now and I'll figure out a better name once I figure out how to enable it :)

**Review by mswiniarsk — APPROVED** [review]
LGTM

**Review by kdamaszk — APPROVED** [review]
